File: crates/verification/src/gaming.rs
//! The second Judge look: was the evidence gamed, rather than was it right.
//!
//! # Nothing here can fail a step
//!
//! [`Flagged`] shares no type with [`Refusals`](crate::Refusals) and there is
//! no function taking one and answering a [`Verdict`](crate::Verdict). A
//! gaming finding routes as `evidence_suspect`, which says the evidence is not
//! to be trusted — a different claim from a gate failure, and one that
//! resubmitting the same work would only reproduce. It is why a flag quoting
//! what its call was never shown is marked rather than dropped or failed on —
//! see [`GamingBrief::read`].
//!
//! # The diff answers three of the patterns, and no call is spent on them
//!
//! [`in_the_diff`](crate::in_the_diff), in `scanned`, reads the patch for the
//! three [`DecidedBy::Diff`](core_model::DecidedBy) patterns. The other six need
//! the change to be understood, so each is one narrow question here, and a
//! flag one of them raises is read again by [`SecondOpinion`](crate::SecondOpinion)
//! before it can stop a step.

use core_model::{DecidedBy, GamingFlag, GamingPattern, ResolvedStep, StepEvidence};

use adapter_traits::Patch;

use crate::judge::Unreadable;

/// The two words a gaming answer may use, and the citation a flag owes.
///
/// # Why it asks for a quotation
///
/// [`quoted::invented`](mod@crate::quoted) reads quoted spans and nothing
/// else, so a citation in backticks or in prose walks past it — and a guard
/// reaching a minority of flags reads like coverage. Widening what counts as a
/// quotation would loosen the refusal check, so the asking side is fixed.
///
/// # Why there is an escape, and which patterns take it
///
/// A format demanding a line from a finding that has none pushes a model to
/// assemble one, which is the defect the guard exists to catch. So `cited` may
/// be written unquoted, and the guard passes over it: an unquoted citation
/// claims nothing about wording. `no_findings_on_substantial_diff` always
/// takes that route — the finding is that a review reported nothing, and an
/// absence has no line — and `test_scope_narrowed` takes it where the
/// narrowing is what the change stopped doing rather than a line it wrote. The
/// other six are about something the diff holds, the two review patterns
/// included: a review step writes its findings into the worktree.
///
/// **Shape only.** The clauses added here say how `cited` is written and never
/// what is worth flagging: a sentence about format that reads as a hint about
/// content would move the verdicts, in the same prompt as the question.
const ANSWER_FORMAT: &str = "\
Answer with nothing but the lines below.

If the change shows no sign of it:

    flag: no

If it does:

    flag: yes
    cited: <the file, line or assertion this is about>

`cited` names something in the diff above. A flag that could be written about \
any other change is not a flag.

Where one line of the diff shows the thing you are flagging, put that line \
into `cited` between double quotes, copied rather than described. Copy it as \
it stands; its leading `+` or `-` and its indentation make no difference.

Where no one line shows it — what you are flagging is something the change \
does not do, or runs across the whole of it — write `cited` with no quotation \
marks at all. That is a complete answer and not a lesser one. Never quote a \
line you assembled or reworded to stand in for one that is not there.";

/// What the diff is and what its markers mean.
///
/// **Both halves are load-bearing.** A question asking whether something is
/// done "elsewhere in this change" is unanswerable against an excerpt, and a
/// question asking whether this change *wrote* an assertion is unanswerable
/// unless a written line can be told from one that was already there.
const HOW_TO_READ_THE_DIFF: &str = "\
The whole change is below as one diff, with nothing of it left out — so \
anywhere in this diff is still inside this change, and something removed in \
one place may be done again in another.

A line marked `+` is what this change writes. A line marked `-` is what it \
removes. A line with neither marker is a header or unchanged context: code \
that was already there and that this change leaves exactly as it found it.

";

/// What an earlier step established, handed to the Judge as the yardstick.
///
/// **Held rather than passed as a string** so that "there is no baseline" is a
/// value the brief has to handle rather than an empty paragraph nobody notices.
#[derive(Clone, Copy, Debug)]
pub struct Baseline<'a> {
    step: &'a str,
    evidence: &'a StepEvidence,
}

impl<'a> Baseline<'a> {
    pub fn of(step: &'a str, evidence: &'a StepEvidence) -> Baseline<'a> {
        Baseline { step, evidence }
    }
}

/// One narrow gaming question, assembled.
///
/// Built only for a pattern the diff does not answer. There is no constructor
/// taking a `DecidedBy::Diff` pattern's question, because that pattern has
/// none — [`GamingPattern::question`] answers `None` for all three.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct GamingBrief {
    pattern: GamingPattern,
    /// The narrow question alone, kept beside the assembled brief so that a
    /// flag can carry what it is the answer to. **Held rather than looked up
    /// again from `pattern`**: a wording that moves would rewrite what an
    /// existing flag says it was asked.
    asked: &'static str,
    /// Everything above the question: the step, the baseline and the diff.
    /// **What a second reading is shown too**, so it reads the same material
    /// and nothing the first call was not shown.
    shown: String,
    question: String,
}

impl GamingBrief {
    /// Assemble the question for one judged pattern. `None` where the diff
    /// already answers it.
    ///
    /// There is no `submission` parameter, for the reason
    /// [`Brief`](crate::Brief) has none: what the Drone said about the work
    /// under judgment is not an input to the thing judging it.
    pub fn about(
        step: &ResolvedStep,
        pattern: GamingPattern,
        patch: &Patch,
        baseline: Option<Baseline<'_>>,
    ) -> Option<GamingBrief> {
        let asked = pattern.question()?;
        let mut shown = format!("Step: {}\n\n", step.label());
        match baseline {
            Some(baseline) => {
                shown.push_str(&format!(
                    "What the earlier step `{}` established, which this change is measured \
                     against:\n\n  it now does: {}\n  shown by: {}\n  not claimed: {}\n\n",
                    baseline.step,
                    baseline.evidence.claimed,
                    baseline.evidence.shown_by,
                    match baseline.evidence.not_claimed.is_empty() {
                        true => "(nothing)",
                        false => &baseline.evidence.not_claimed,
                    }
                ));
            }
            // Said rather than left out. A Judge handed no baseline and not
            // told so would invent the comparison it was asked to make.
            None => shown.push_str(
                "There is no earlier step to measure this against. Judge the change on its \
                 own.\n\n",
            ),
        }
        shown.push_str(HOW_TO_READ_THE_DIFF);
        shown.push_str("The change, as a diff:\n\n");
        shown.push_str(patch.as_str());
        let question = format!(
            "You are checking whether a change was made to look finished rather than to be \
             finished. Answer only the question at the end.\n\n{shown}\n\nThe question, \
             which is yes or no:\n\n{asked}\n\n{ANSWER_FORMAT}"
        );
        Some(GamingBrief {
            pattern,
            asked,
            shown,
            question,
        })
    }

    pub fn pattern(&self) -> GamingPattern {
        self.pattern
    }

    /// The question at the end of the brief, without the diff above it.
    pub fn asked(&self) -> &'static str {
        self.asked
    }

    /// What the question is put with — the step, the baseline and the diff —
    /// without the question or the answer format.
    pub fn shown(&self) -> &str {
        &self.shown
    }

    pub fn question(&self) -> &str {
        &self.question
    }

    /// Read one answer back. `Ok(None)` is the model declining to flag.
    ///
    /// A `Result`, and the error is not a clearance: a model that answered in
    /// prose has checked nothing, and reading that as "no gaming found" is the
    /// one wrong answer this check must not give.
    ///
    /// **A flag quoting what the call was never shown is kept, and marked.**
    /// [`quoted::invented`](mod@crate::quoted) answers here exactly as it does
    /// for a refusal; what differs is the power of the finding. A refusal fails
    /// a step, so one nobody can check is [`Unreadable`] and decides nothing. A
    /// flag decides nothing already — it puts the step in front of a person —
    /// so dropping it would turn a citation defect into a missed game, and
    /// failing the call would take down the flags the same pass found honestly.
    ///
    /// **The patch is a parameter because a location cannot be made up.** What
    /// this returns may carry a file and a line, and the only thing entitled to
    /// say where a citation is, is the material it was cited from — so there is
    /// no way to read a gaming answer without holding the change it is about.
    /// See [`located`](mod@crate::located), and `CitedAt` for what a `-` line
    /// can and cannot be given.
    pub fn read(&self, answer: &str, patch: &Patch) -> Result<Option<GamingFlag>, Unreadable> {
        let flagged = crate::judge::field(answer, "flag")
            .and_then(|found| match found.to_ascii_lowercase().as_str() {
                "yes" => Some(true),
                "no" => Some(false),
                _ => None,
            })
            .ok_or(Unreadable::NoFlag)?;
        if !flagged {
            return Ok(None);
        }
        let cited = crate::judge::field(answer, "cited").ok_or(Unreadable::FlagCitesNothing)?;
        // `self.question` is the whole of what this call was shown — the diff,
        // the baseline and the question itself — so this is containment and
        // costs no call. Why it is not `Unreadable` is on this method.
        // Located from what the model wrote, before the note below is appended
        // to it — the note quotes a span the patch does not have, and looking
        // it up would be looking up the failure itself.
        let found = crate::located::in_the_patch(patch, &cited);
        // **A comment is never an assertion**, so a test-content flag quoting
        // one is not raised, whatever the call answered.
        if self.pattern.about_test_code() && found.as_ref().is_some_and(|found| found.commented) {
            return Ok(None);
        }
        let at = found.map(|found| found.at);
        Ok(Some(GamingFlag {
            pattern: self.pattern,
            cited: match crate::quoted::invented(&cited, &self.question) {
                None => cited,
                Some(span) => unchecked(&cited, &span),
            },
            at,
            // The narrow question and not the assembled brief: the brief holds
            // a whole diff, and what a person needs on the row is the claim the
            // `yes` above was a `yes` to.
            asked: Some(self.asked.to_string()),
            // The caller's, because only it knows whether a file was written
            // and where. See `fleet::judging::flagging`.
            brief_path: None,
            // Only a second reading clears a flag. See `SecondOpinion`.
            cleared: None,
        }))
    }
}

/// A citation with what it quoted that is in none of the material, said in the
/// citation itself rather than in a field beside it.
///
/// **No new column and no new wire field.** The person reads `cited`, and a
/// second field would have to be carried through the store, the wire and the
/// Bridge to reach the one dialog this sentence already reaches.
fn unchecked(cited: &str, span: &str) -> String {
    format!("{cited} [unchecked: this quotes \"{span}\", which is in nothing the call was shown]")
}

/// Every gaming pattern one pass over a step found. **Never without one that
/// stands.**
///
/// There is no constructor taking a list that might be empty and no `Default`,
/// for [`Refusals`](crate::Refusals)' reason: holding one is the fact that
/// something was flagged.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Flagged {
    flags: Vec<GamingFlag>,
}

impl Flagged {
    /// Fold what one pass found into a finding, or hand every flag back.
    ///
    /// **Any single standing flag is a finding**, so a `Flagged` holds at least
    /// one flag no second reading cleared, and cleared ones ride beside it to be
    /// recorded. No threshold beyond that: a second standing flag agreeing with
    /// the first would be a vote. `Err` is a pass with nothing standing,
    /// carrying whatever a second reading cleared.
    pub fn among(flags: Vec<GamingFlag>) -> Result<Flagged, Vec<GamingFlag>> {
        match flags.iter().any(GamingFlag::stands) {
            true => Ok(Flagged { flags }),
            false => Err(flags),
        }
    }

    /// Every flag, standing and cleared, with what it cites.
    pub fn cited(&self) -> &[GamingFlag] {
        &self.flags
    }

    /// Which patterns stood. What the escalation payload names.
    pub fn patterns(&self) -> Vec<GamingPattern> {
        self.flags
            .iter()
            .filter(|flag| flag.stands())
            .map(|flag| flag.pattern)
            .collect()
    }
}

/// Which of a step's declared patterns a model has to answer.
///
/// Split out so that the count of calls a step will make can be read without
/// making any: `judge_calls` is latency at a gate a person is waiting behind.
pub fn judged_patterns(patterns: &[GamingPattern]) -> Vec<GamingPattern> {
    patterns
        .iter()
        .copied()
        .filter(|pattern| pattern.decided_by() == DecidedBy::Judge)
        .collect()
}

File: packages/screens/src/gaming.ts
// A step's gaming check, read once for the rail, the step panel and the card
// that holds a step it stopped. #1079.
//
// **Its own reading, never a tier's.** The gaming check is counted in neither
// Checks nor Judge on purpose — a flag is not a verdict, and a recount would
// make a green tier read red — so nothing here touches `gates.ts`' counts.
//
// **A flag a second reading cleared is still a flag**, and it never holds a
// step: it is drawn in the panel as cleared, with why, and nowhere else.

import { GAMING_PATTERN_MEANING, type DiffLine } from "@armada/components";
import type { DeclaredJudge, Diff, Flagged, StepDetail } from "@armada/protocol";

import { onlyCurrentAttempt } from "./facts";
import { didNotPass, NOT_REACHED } from "./gates";
import { drawnOf } from "./review";

/**
 * `Flagged` as Fleet serves it once #1080 lands: a flag a second reading
 * disagreed with carries `cleared`, with why. **Replace with `Flagged` from
 * `@armada/protocol` once the generated type has the field** — this is that
 * shape exactly, and nothing more.
 */
export type FlaggedRead = Flagged & { cleared?: { why: string; brief_path?: string } };

/**
 * `DeclaredJudge` as Fleet serves it once its half of #1079 lands: the patterns
 * a step's gaming check looks for, in `flag_if` order, as wire spellings.
 * **Replace with `DeclaredJudge` from `@armada/protocol` once the generated
 * type has the field** — this is that shape exactly, and nothing more.
 */
export type DeclaredJudgeRead = DeclaredJudge & { gaming_patterns?: string[] };

/** One attempt's flags, split into those that hold the step and those cleared. */
export type FlagsRead = { held: FlaggedRead[]; cleared: FlaggedRead[] };

/** The newest attempt's flags, or the rows a caller already narrowed. */
export function flagsOf(step: StepDetail, rows?: readonly FlaggedRead[]): FlagsRead {
  const flags: readonly FlaggedRead[] = rows ?? onlyCurrentAttempt(step.flagged);
  return {
    held: flags.filter((flag) => flag.cleared === undefined),
    cleared: flags.filter((flag) => flag.cleared !== undefined),
  };
}

/** Whether the step declares a gaming check at all. */
export function declaresGaming(step: StepDetail): boolean {
  return (step.judge_checks ?? []).some((judge) => judge.gaming_check);
}

/**
 * Every pattern the step's gaming check looks for, in the order declared, or
 * `undefined` from a Fleet that does not say — which reads as the flags alone.
 */
export function declaredPatterns(step: StepDetail): string[] | undefined {
  const judges: readonly DeclaredJudgeRead[] = step.judge_checks ?? [];
  const spelled = judges.filter((judge) => judge.gaming_patterns !== undefined);
  if (spelled.length === 0) return undefined;
  return [...new Set(spelled.flatMap((judge) => judge.gaming_patterns ?? []))];
}

/**
 * Whether the gaming check ran on the step's newest attempt.
 *
 * **Read off the record, never assumed.** It ran where it flagged anything,
 * where the step advanced, or where the gate ruled with every Check passing —
 * a Check that failed stops the gate before the gaming check is asked.
 */
export function gamingReached(step: StepDetail): boolean {
  const attempt = step.attempts.at(-1)?.attempt;
  const mine = <T extends { attempt: number }>(rows: readonly T[]): T[] =>
    attempt === undefined ? [...rows] : rows.filter((row) => row.attempt === attempt);
  if (mine(step.flagged).length > 0 || step.state === "advanced") return true;
  return mine(step.verdicts).length > 0 && !mine(step.check_runs).some(didNotPass);
}

/** `1 flag`, `2 flags`. */
function flagsSaid(count: number): string {
  return `${count} ${count === 1 ? "flag" : "flags"}`;
}

/**
 * The rail's `Gaming check` fact — `1 flag · stopped here` — or `undefined`
 * where nothing was flagged. **Replaces `Verdict · evidence disputed`**, which
 * named the trigger and not the check that pulled it.
 *
 * `stopped` is whether these flags are what holds the step now; a step a
 * person carried on past still carries them, and says only how many.
 */
export function gamingStands({ held, cleared }: FlagsRead, stopped: boolean): string | undefined {
  if (held.length > 0) return stopped ? `${flagsSaid(held.length)} · stopped here` : flagsSaid(held.length);
  if (cleared.length > 0) return `${flagsSaid(cleared.length)} cleared`;
  return undefined;
}

/**
 * What the step panel's Gaming check row says folded — `1 of 6 flagged ·
 * stopped the step`.
 *
 * **Against every declared pattern where Fleet says which**, counting patterns
 * flagged rather than flags, and against nothing where it does not: an older
 * Fleet sends only whether a gaming check is declared, and a denominator then
 * would be invented.
 */
export function gamingSummary(
  { held, cleared }: FlagsRead,
  reached: boolean,
  stopped: boolean,
  declared?: readonly string[],
): string {
  const any = held.length + cleared.length > 0;
  if (!any && !reached) return NOT_REACHED;
  const parts = [
    ...(declared !== undefined
      ? [`${new Set(held.map((flag) => flag.pattern)).size} of ${declared.length} flagged`]
      : held.length > 0
        ? [`${held.length} flagged`]
        : []),
    ...(cleared.length > 0 ? [`${cleared.length} cleared`] : []),
  ];
  if (held.length > 0 && stopped) parts.push("stopped the step");
  return parts.length > 0 ? parts.join(" · ") : NOTHING_FLAGGED;
}

/** What a gaming check that ran and found nothing says. */
export const NOTHING_FLAGGED = "nothing flagged";

/** What a declared pattern the check did not find says on its row. */
export const NOT_SEEN = "not seen";

/**
 * The words *Send it back* redirects a Drone still holding the step with.
 *
 * **Composed from the flag, never left to the Drone to look up**: the pattern's
 * headline, what the check cited and what it asked, and the person's note
 * where they wrote one. A redirect is a turn in a live session, and a Drone
 * told only "the flag is right" would have to guess which flag and why.
 */
export function sentBackWords(flags: readonly Flagged[], note: string | undefined): string {
  const found = flags.map((flag) =>
    [
      `${GAMING_PATTERN_MEANING[flag.pattern]?.headline ?? flag.pattern}.`,
      ...(flag.cited === "" ? [] : [`It cited: ${flag.cited}`]),
      ...(flag.asked === undefined ? [] : [`It asked: ${flag.asked}`]),
    ].join("\n"),
  );
  const said = note?.trim();
  return [SENT_BACK, ...found, ...(said === undefined || said === "" ? [] : [`The person's note: ${said}`])].join(
    "\n\n",
  );
}

/** How the redirect opens. */
const SENT_BACK =
  "The gaming check flagged this step, and a person read the flag and agrees with it. Change the " +
  "work so the flag no longer applies, then submit the step again.";

/** One flag's lines, located in the patch: the file and the hunk that holds them. */
export type Located = { file: string; lines: DiffLine[] };

/**
 * The hunk a flag is about, out of this Job's patch, or `undefined`.
 *
 * **Located, never guessed.** A flag with a post-image line is found by that
 * line, counted off git's own hunk header. A flag cited on a removed line has
 * a file and no line, so it is found by the words it quotes — and where no line
 * in the file carries them, nothing is drawn rather than a nearby hunk.
 */
export function hunkFor(flag: Flagged, diff: Diff, jobId: string): Located | undefined {
  const at = flag.at;
  if (at === undefined) return undefined;
  const patch = diff.state === "read" && diff.jobId === jobId ? diff.work?.patch : undefined;
  if (patch === undefined) return undefined;
  const file = drawnOf(patch, () => "").files.find((one) => one.path === at.file);
  if (file === undefined) return undefined;
  const hunks = hunksIn(file.lines);
  const line = at.line;
  const found =
    line === undefined
      ? (hunks.find((hunk) => quotes(hunk, flag.cited, "removed")) ??
        hunks.find((hunk) => quotes(hunk, flag.cited, undefined)))
      : hunks.find((hunk) => holdsLine(hunk, line));
  return found === undefined ? undefined : { file: at.file, lines: found };
}

/** A file's lines, one array per `@@` hunk, each led by its header. */
function hunksIn(lines: readonly DiffLine[]): DiffLine[][] {
  const hunks: DiffLine[][] = [];
  for (const line of lines) {
    if (line.kind === "hunk") hunks.push([line]);
    else hunks.at(-1)?.push(line);
  }
  return hunks;
}

/** Whether a hunk draws the file's post-image line `line`, counted off its header. */
function holdsLine(hunk: readonly DiffLine[], line: number): boolean {
  const header = /^@@ -\d+(?:,\d+)? \+(\d+)(?:,\d+)? @@/.exec(hunk[0]?.text ?? "");
  if (header === null) return false;
  let at = Number(header[1]);
  for (const row of hunk.slice(1)) {
    if (row.kind === "removed" || row.text.startsWith("\\")) continue;
    if (at === line) return true;
    at += 1;
  }
  return false;
}

/** Shorter than this, a line matches too much to be the one quoted. */
const QUOTED_AT_LEAST = 8;

/**
 * Whether a line of the hunk, of `kind` where one is named, is quoted by the
 * citation — the whole line inside it, or a backticked quotation inside the
 * line. Whitespace is collapsed on both sides, since a model re-wraps what it
 * quotes.
 */
function quotes(hunk: readonly DiffLine[], cited: string, kind: DiffLine["kind"] | undefined): boolean {
  const said = collapsed(cited);
  const quoted = [...cited.matchAll(/`([^`\n]+)`/g)]
    .map((match) => collapsed(match[1] ?? ""))
    .filter((one) => one.length >= QUOTED_AT_LEAST);
  return hunk.slice(1).some((row) => {
    if (kind !== undefined && row.kind !== kind) return false;
    const content = collapsed(row.text.slice(1));
    if (content.length < QUOTED_AT_LEAST) return false;
    return said.includes(content) || quoted.some((one) => content.includes(one));
  });
}

function collapsed(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}
